File: DataStructure/LinkedList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace C_sharp_DataStructure_Sort
{
    /// <summary>
    /// Author:Jigger
    /// Date:2022/1
    /// 双端链表
    /// </summary>
    //双端链表节点
    class LinkedNode<T>
    {
        private T _data;
        public T Data
        {
            get => _data;
            set => _data = value;
        }
        public LinkedNode<T> NextNode;
        public LinkedNode()
        {
            NextNode = null;
        }
        public LinkedNode(T data)
        {
            _data = data;
            NextNode = null;
        }
    }
    //双端链表
    class LinkedList<T>
    {
        private LinkedNode<T> _head;//头节点，不计入链表长度，不作为存放数据的节点
        private LinkedNode<T> _tail;//尾节点，最后一个数据节点
        private int _length = 0;
        public int Length { get => _length; }
        public LinkedNode<T> Head { get => _head; set => _head = value; }
        public LinkedNode<T> Tail { get => _tail; set => _tail = value; }

        public LinkedList() 
        {
            Head = new LinkedNode<T>();
            Tail = Head;
        }
        //构造定长链表
        public LinkedList(int len) 
        {
            _length = len;
            int i = 0;
            Head = new LinkedNode<T>();
            LinkedNode<T> cur = Head;
            while (i < len)
            {
                cur.NextNode = new LinkedNode<T>();
                cur = cur.NextNode;
                i++;
            }
            Tail = cur;
        }
        //判空
        public bool IsEmpty() 
        {
            if (Head == Tail)
            {
                return true;
            }
            return false;
        }
        //尾插
        public void PushBack(LinkedNode<T> newNode) 
        {
            Tail.NextNode = newNode;
            Tail = newNode;
            _length++;
        }
        //头插
        public void PushFront(LinkedNode<T> newNode)
        {
            newNode.NextNode = Head.NextNode;
            Head.NextNode = newNode;
            _length++;
        }
        //尾删,并弹出(即Append方法)
        public LinkedNode<T> PopBack()
        {
            if (IsEmpty() || HasCircle() != null) 
            {
                return null;
            }
            LinkedNode<T> cur = Head;
            while (cur.NextNode != Tail) 
            {
                cur = cur.NextNode;    
            }           
            Tail = cur;
            _length--;
            return cur.NextNode;
        }
        //头删，并弹出
        public LinkedNode<T> PopFront() 
        {
            if (IsEmpty())
            {
                return null;
            }
            LinkedNode<T> cur = Head;
            Head = cur.NextNode;
            _length--;
            return cur;
        }
        //删除索引位置的节点（从0开始，0为第一个）
        public void Delete(int index)
        {
            if (index >= Length || HasCircle() != null) 
            {
                Console.WriteLine("错误！");
                return;
            }
            LinkedNode<T> pre = Head;
            int i = 0;
            while (i < index)
            {
                pre = pre.NextNode;
                i++;
            }
            pre.NextNode = pre.NextNode.NextNode;
            _length--;
        }
        //清空
        public void Clear() 
        {
            Head.NextNode = null;
            Tail = Head;
            _length = 0;
        }
        //查找
        public LinkedNode<T> Find(T data) 
        {
            if (IsEmpty() || HasCircle() != null)
            {
                return null;
            }
            LinkedNode<T> cur = Head;
            while (cur != Tail && !cur.Data.Equals(data))
            {
                cur = cur.NextNode;
            }
            if (cur.Data.Equals(data)) 
            {
                return cur;
            }
            return null;
        }
        //查找索引
        public int FindIndex(T data)
        {
            int i = -1;
            if (IsEmpty() || HasCircle() != null)
            {
                return -1;
            }
            LinkedNode<T> cur = Head;
            while (cur != Tail && !cur.Data.Equals(data))
            {
                cur = cur.NextNode;
                i++;
            }
            if (cur.Data.Equals(data))            
            {
                return i;
            }
            return -1;
        }
        //查找索引处的值
        public T FindData(int index) 
        {
            if (index >= Length || HasCircle() != null)
            {
                Console.WriteLine("错误！");
                return default(T);
            }
            LinkedNode<T> pre = Head;
            int i = -1;
            while (i < index)
            {
                pre = pre.NextNode;
                i++;
            }
            return pre.Data;
        }
        //索引插入
        public void Insert(int index, LinkedNode<T> newNode)
        {
            if (index >= Length || HasCircle() != null)
            {
                Console.WriteLine("越界！");
                return;
            }
            LinkedNode<T> pre = Head;
            int i = -1;
            while (i < index-1)
            {
                pre = pre.NextNode;
                i++;
            }
            newNode.NextNode = pre.NextNode;
            pre.NextNode = newNode;
            _length++;
        }
        //节点位置插入
        public bool Insert(LinkedNode<T> insertNode, LinkedNode<T> newNode) 
        {
            if (IsEmpty() || HasCircle() != null)
            {
                return false;
            }
            LinkedNode<T> cur = Head;
            while (cur.NextNode != Tail && cur.NextNode != insertNode)
            {
                cur = cur.NextNode;
            }
            if (cur.NextNode == insertNode)
            {
                newNode.NextNode = cur.NextNode;
                cur.NextNode = newNode;
                return true;
            }
            _length++;
            return false;
        }
        //反转链表
        public void Reverse() 
        {
            LinkedNode<T> p = Head.NextNode, q = null, r = Head;
            Tail = Head.NextNode;
            Head = Head.NextNode;
            while (p != null) 
            {
                p = p.NextNode;
                Head.NextNode = q;
                q = Head;
                Head = p;
            }
            Head = r;
            Head.NextNode = q;
        }
        //判断是否有环并返回入口节点
        public LinkedNode<T> HasCircle() 
        {
            List<LinkedNode<T>> tag = new();
            tag.Add(Head);
            LinkedNode<T> cur = Head;
            while (cur != null && cur.NextNode != null) 
            {
                if (tag.Contains(cur.NextNode)) 
                {
                    return cur;
                }
                tag.Add(cur.NextNode);
                cur = cur.NextNode;
            }
            return null;
        }
    }
}

File: DataStructure/Linkedlist.lua
-- Author: Jigger
-- Date: 2021-12
-- 功能描述:单链表Linkedlist

Linkedlist = {}

-- 初始化，构建一个空表
function Linkedlist:New()
    self.__index = self
    self.length = 0
    self.next = nil
    print("链表初始化成功")
    return setmetatable({},self)
end

function Linkedlist:CheckList()
    if not self then print("链表不存在") return end
end

function Linkedlist:CheckIndex(index)
    if index < 1 or index > self.length + 1 then
        print("索引越界")
        return
    end
end

function Linkedlist:CheckAll(index)
    Linkedlist:CheckList()
    Linkedlist:CheckIndex(index)
end

-- 尾部添加节点
function Linkedlist:AddTail(data)
    Linkedlist:CheckList()
    local node = self
    while node.next do
        node = node.next
    end
    -- 创建新的尾部节点
    node.next = {}
    node = node.next
    node.next = nil
    node.data = data
    self.length = self.length+1
    print("尾部添加节点成功")
end

-- 头部添加节点
function Linkedlist:AddHead(data)
    Linkedlist:CheckList()
    --修改新节点和头节点的next
    local newNode = {}
    newNode.data = data
    newNode.next = self.next
    self.next = newNode
    self.length = self.length+1
    print("头部添加节点成功")
end

-- 索引添加节点
function Linkedlist:InsertByIndex(index, data)
    Linkedlist:CheckAll(index)
    local j,k,l = index - 1, 0, self
    while k ~= j do
        k = k + 1
        l = l.next
    end
    -- 开始添加
    local newNode = {}
    newNode.data = data
    newNode.next = l.next
    l.next = newNode
    self.length = self.length + 1
    print("索引添加节点成功")
end

-- 索引删除并返回节点数据内容
function Linkedlist:DeleteByIndex(index)
    Linkedlist:CheckAll(index)
    local j,k,l = index - 1, 0, self
    while k ~= j do
        k = k + 1
        l = l.next
    end
    -- 开始删除 l.next
    d = l.next.data
    t = l.next.next
    l.next = nil
    l.next = t
    self.length = self.length - 1
    print("索引删除节点成功")
end

-- 索引修改节点内容
function Linkedlist:ModifyDataByIndex(index, data)
    Linkedlist:CheckAll(index)
    local l = self.next
    local k = 1
    while l do
        if k == index then
            l.data = data
            print("索引修改节点成功")
            return
        end
        l = l.next
        k = k + 1
    end
    print("索引越界")
end

-- 索引获取指定节点内容
function Linkedlist:GetDataByIndex(index)
    Linkedlist:CheckAll(index)
    local l = self.next
    local k = 1
    while l do
        if k == index then
            print("索引获取节点成功")
            return l.data
        end
        l = l.next
        k = k + 1
    end
    print("索引越界")
end

-- 反转倒置链表
function Linkedlist:Reverse()
    Linkedlist:CheckList()
    -- p为当前遍历的节点 q为前遍历节点的前一节点 pr中间变量
    -- ret 返回反转后的链表
    local p, q= self.next, nil
    self.next = nil
    while p do
        pr = p.next
        p.next = q
        q = p
        p = pr
    end
    self.next = q
    return self
end

-- 打印链表元素
function Linkedlist:Display()
    Linkedlist:CheckList()
    local l = self.next
    local x = 1
    while l do
        print(x .. " : " .. l.data)
        l = l.next
        x = x + 1
    end
end

-- 链表是否为空
function Linkedlist:IsEmpty()
    Linkedlist:CheckList()
    return self.length == 0
end

-- 清理链表，操作完成后，链表还在，只不过为空，相当刚开始的初始化状态
function Linkedlist:Clear()
    Linkedlist:CheckList()
    while true do
        -- fNode为遍历用指针
        fNode = self.next
        -- 满足则已为空表
        if not fNode then
            print("链表清理成功")
            break
        end
        t = fNode.next
        -- 置空
        self.next = nil
        self.next = t
    end
    list.length = 0
end

-- 销毁链表
function Linkedlist:Destory()
    if self then
        Linkedlist:Clear()
    end
    self = nil
    print("链表销毁成功")
end

return Linkedlist